File: backend/app/schemas/reliability.py
"""Pydantic schemas for Unified Reliability, SRE, Disaster Recovery, and Operations Platform."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


# --- Health & Diagnostics ---

class ComponentHealthCheckSchema(BaseModel):
    component_name: str
    component_type: str
    status: str
    latency_ms: float
    is_critical: bool = True
    message: Optional[str] = None
    details: Optional[Dict[str, Any]] = None
    checked_at: Optional[str] = None


class DeepHealthResultSchema(BaseModel):
    status: str
    active_circuit_breakers: int = 0
    open_incidents: int = 0
    components: Dict[str, ComponentHealthCheckSchema]
    timestamp: str


class LivenessResponse(BaseModel):
    status: str = "ALIVE"
    timestamp: str


class ReadinessResponse(BaseModel):
    status: str = "READY"
    timestamp: str
    checks: Dict[str, bool]


# --- Circuit Breakers ---

class CircuitBreakerStatusSchema(BaseModel):
    service_name: str
    state: str
    failure_count: int
    success_count: int
    failure_threshold: int
    recovery_timeout: float
    last_state_change: Optional[str] = None


class CircuitBreakerResetRequest(BaseModel):
    service_name: str


# --- SLO & Error Budgets ---

class SLOCreateSchema(BaseModel):
    slo_type: str
    name: str
    description: Optional[str] = None
    target_percentage: float = Field(default=99.9, ge=0.0, le=100.0)
    window_days: int = Field(default=30, gt=0)
    service_tier: str = "CRITICAL"


class SLOMetricSnapshotSchema(BaseModel):
    slo_type: str
    name: str
    target_slo: float
    current_sli: float
    error_budget_remaining_pct: float
    burn_rate_1h: float
    burn_rate_24h: float
    budget_status: str
    total_events: int
    bad_events: int
    recorded_at: str


class ErrorBudgetSummarySchema(BaseModel):
    overall_status: str
    slos: List[SLOMetricSnapshotSchema]
    generated_at: str


# --- Incidents & Postmortems ---

class IncidentCreateSchema(BaseModel):
    title: str
    severity: str
    affected_services: List[str]
    impact_summary: str
    lead_responder: Optional[str] = None
    responders: List[str] = Field(default_factory=list)


class IncidentUpdateSchema(BaseModel):
    title: Optional[str] = None
    severity: Optional[str] = None
    status: Optional[str] = None
    impact_summary: Optional[str] = None
    lead_responder: Optional[str] = None
    responders: Optional[List[str]] = None


class IncidentTimelineEventAddSchema(BaseModel):
    description: str
    actor: str


class IncidentMitigationStepAddSchema(BaseModel):
    step: str
    completed: bool = True


class IncidentResponseSchema(BaseModel):
    id: str
    title: str
    severity: str
    status: str
    affected_services: List[str]
    lead_responder: Optional[str] = None
    responders: List[str]
    impact_summary: str
    timeline_events: List[Dict[str, Any]]
    mitigation_steps: List[Dict[str, Any]]
    started_at: str
    acknowledged_at: Optional[str] = None
    mitigated_at: Optional[str] = None
    resolved_at: Optional[str] = None
    created_at: Optional[str] = None


class PostmortemCreateSchema(BaseModel):
    incident_id: str
    summary: str
    root_cause: str
    five_whys: List[str] = Field(default_factory=list)
    what_went_well: List[str] = Field(default_factory=list)
    what_could_improve: List[str] = Field(default_factory=list)
    action_items: List[Dict[str, Any]] = Field(default_factory=list)
    owner: str


class PostmortemResponseSchema(BaseModel):
    id: str
    incident_id: str
    summary: str
    root_cause: str
    five_whys: List[str]
    timeline_events: List[Dict[str, Any]]
    what_went_well: List[str]
    what_could_improve: List[str]
    action_items: List[Dict[str, Any]]
    owner: str
    created_at: str


# --- Backups & Restore Verification ---

class BackupTriggerSchema(BaseModel):
    backup_type: str = "DATABASE_FULL"
    storage_location: Optional[str] = None
    retention_days: int = 30


class BackupRecordSchema(BaseModel):
    id: str
    backup_type: str
    status: str
    size_bytes: int
    storage_location: str
    checksum_sha256: str
    verified: bool
    retention_days: int
    started_at: str
    completed_at: Optional[str] = None


class RestoreVerificationTriggerSchema(BaseModel):
    backup_id: str
    environment: str = "ISOLATED_SANDBOX"


class RestoreVerificationResponseSchema(BaseModel):
    test_id: str
    backup_id: str
    environment: str
    status: str
    rto_achieved_seconds: int
    data_integrity_passed: bool
    tables_restored_count: int
    records_verified_count: int
    logs: List[Dict[str, Any]]
    executed_at: str


# --- Disaster Recovery Plans & Drills ---

class DRPlanSchema(BaseModel):
    plan_name: str
    scenario: str
    target_rpo_minutes: int
    target_rto_minutes: int
    primary_region: str
    secondary_region: str
    status: str
    recovery_steps: List[Dict[str, Any]]


class DRDrillTriggerSchema(BaseModel):
    scenario: str
    initiated_by: str
    plan_name: Optional[str] = None


class DRDrillResponseSchema(BaseModel):
    drill_id: str
    scenario: str
    initiated_by: str
    status: str
    actual_rto_minutes: float
    data_loss_minutes: float
    step_results: List[Dict[str, Any]]
    observations: str
    lessons_learned: List[str]
    started_at: str
    completed_at: str


# --- Operations: Deployments, Rollbacks, Feature Flags ---

class DeploymentRecordCreateSchema(BaseModel):
    version: str
    deployed_by: str
    environment: str = "PRODUCTION"
    git_commit_sha: Optional[str] = None
    release_notes: Optional[str] = None


class DeploymentRecordSchema(BaseModel):
    id: str
    version: str
    environment: str
    deployed_by: str
    git_commit_sha: str
    status: str
    smoke_tests_passed: bool
    migrations_applied: bool
    release_notes: Optional[str] = None
    deployed_at: str


class RollbackTriggerSchema(BaseModel):
    current_version: str
    target_version: str
    error_rate_pct: float
    p99_latency_ms: float
    smoke_tests_passed: bool
    initiated_by: str
    reason: Optional[str] = None


class FeatureFlagCreateUpdateSchema(BaseModel):
    name: str
    description: Optional[str] = None
    enabled: bool = True
    percentage_rollout: int = Field(default=100, ge=0, le=100)
    allowed_tiers: List[str] = Field(default_factory=list)
    tenant_whitelist: List[str] = Field(default_factory=list)
    tenant_blacklist: List[str] = Field(default_factory=list)


class FeatureFlagEvaluationRequest(BaseModel):
    flag_name: str
    tenant_id: Optional[str] = None
    tenant_tier: Optional[str] = None

File: frontend/lib/api/reliability.ts
import { api } from './client';

export interface ComponentHealth {
  component_name: string;
  component_type: string;
  status: 'HEALTHY' | 'DEGRADED' | 'UNHEALTHY' | 'CRITICAL';
  latency_ms: number;
  is_critical: boolean;
  message?: string;
  details?: Record<string, any>;
  checked_at?: string;
}

export interface DeepHealthResult {
  status: 'HEALTHY' | 'DEGRADED' | 'UNHEALTHY' | 'CRITICAL';
  active_circuit_breakers: number;
  open_incidents: number;
  components: Record<string, ComponentHealth>;
  timestamp: string;
}

export interface CircuitBreakerState {
  service_name: string;
  state: 'CLOSED' | 'OPEN' | 'HALF_OPEN';
  failure_count: number;
  success_count: number;
  failure_threshold: number;
  recovery_timeout: number;
  last_state_change?: string;
}

export interface SLOMetricSnapshot {
  slo_type: string;
  name: string;
  target_slo: number;
  current_sli: number;
  error_budget_remaining_pct: number;
  burn_rate_1h: number;
  burn_rate_24h: number;
  budget_status: 'HEALTHY' | 'WARNING' | 'EXHAUSTED' | 'BREACHED';
  total_events: number;
  bad_events: number;
  recorded_at: string;
}

export interface ErrorBudgetSummary {
  overall_status: string;
  slos: SLOMetricSnapshot[];
  generated_at: string;
}

export interface ReliabilityIncident {
  id: string;
  title: string;
  severity: 'SEV1_CRITICAL' | 'SEV2_MAJOR' | 'SEV3_MODERATE' | 'SEV4_LOW';
  status: 'TRIGGERED' | 'ACKNOWLEDGED' | 'INVESTIGATING' | 'MITIGATED' | 'RESOLVED' | 'CLOSED';
  affected_services: string[];
  lead_responder?: string;
  responders: string[];
  impact_summary: string;
  timeline_events: Array<{
    timestamp: string;
    description: string;
    actor: string;
  }>;
  mitigation_steps: Array<{
    step: string;
    completed: boolean;
  }>;
  started_at: string;
  acknowledged_at?: string;
  mitigated_at?: string;
  resolved_at?: string;
  created_at?: string;
}

export interface IncidentPostmortem {
  id: string;
  incident_id: string;
  summary: string;
  root_cause: string;
  five_whys: string[];
  timeline_events: Array<{
    timestamp: string;
    description: string;
    actor: string;
  }>;
  what_went_well: string[];
  what_could_improve: string[];
  action_items: Array<{
    task: string;
    owner: string;
    status: string;
    due_date?: string;
  }>;
  owner: string;
  created_at: string;
}

export interface BackupRecord {
  id: string;
  backup_type: string;
  status: string;
  size_bytes: number;
  storage_location: string;
  checksum_sha256: string;
  verified: boolean;
  retention_days: number;
  started_at: string;
  completed_at?: string;
}

export interface RestoreVerificationTest {
  test_id: string;
  backup_id: string;
  environment: string;
  status: string;
  rto_achieved_seconds: number;
  data_integrity_passed: boolean;
  tables_restored_count: number;
  records_verified_count: number;
  logs: Array<Record<string, any>>;
  executed_at: string;
}

export interface DRPlan {
  plan_name: string;
  scenario: string;
  target_rpo_minutes: number;
  target_rto_minutes: number;
  primary_region: string;
  secondary_region: string;
  status: string;
  recovery_steps: Array<{
    step: number;
    name: string;
    action: string;
    expected_duration_seconds: number;
    automated: boolean;
  }>;
}

export interface DRDrill {
  drill_id: string;
  scenario: string;
  initiated_by: string;
  status: string;
  actual_rto_minutes: number;
  data_loss_minutes: number;
  step_results: Array<Record<string, any>>;
  observations: string;
  lessons_learned: string[];
  started_at: string;
  completed_at: string;
}

export interface DeploymentRecord {
  id: string;
  version: string;
  environment: string;
  deployed_by: string;
  git_commit_sha: string;
  status: string;
  smoke_tests_passed: boolean;
  migrations_applied: boolean;
  release_notes?: string;
  deployed_at: string;
}

export interface FeatureFlag {
  name: string;
  description?: string;
  enabled: boolean;
  percentage_rollout: number;
  allowed_tiers?: string[];
  tenant_whitelist?: string[];
  tenant_blacklist?: string[];
}

export const reliabilityApi = {
  // Health & Circuits
  getDeepHealth: () => api.get<DeepHealthResult>('/reliability/health'),
  listCircuits: () => api.get<CircuitBreakerState[]>('/reliability/circuits'),
  resetCircuit: (serviceName: string) => api.post('/reliability/circuits/reset', { service_name: serviceName }),

  // SLOs
  getSLODashboard: () => api.get<ErrorBudgetSummary>('/reliability/slos'),
  createSLO: (data: Partial<SLOMetricSnapshot>) => api.post('/reliability/slos', data),

  // Incidents & Postmortems
  listIncidents: (status?: string) => api.get<ReliabilityIncident[]>('/reliability/incidents', { params: { status } }),
  getIncident: (id: string) => api.get<ReliabilityIncident>(`/reliability/incidents/${id}`),
  declareIncident: (data: Partial<ReliabilityIncident>) => api.post<ReliabilityIncident>('/reliability/incidents', data),
  acknowledgeIncident: (id: string) => api.post<ReliabilityIncident>(`/reliability/incidents/${id}/acknowledge`),
  mitigateIncident: (id: string) => api.post<ReliabilityIncident>(`/reliability/incidents/${id}/mitigate`),
  resolveIncident: (id: string) => api.post<ReliabilityIncident>(`/reliability/incidents/${id}/resolve`),
  addIncidentTimeline: (id: string, description: string, actor: string) =>
    api.post<ReliabilityIncident>(`/reliability/incidents/${id}/timeline`, { description, actor }),
  createPostmortem: (data: Partial<IncidentPostmortem>) => api.post<IncidentPostmortem>('/reliability/postmortems', data),
  getPostmortem: (incidentId: string) => api.get<IncidentPostmortem>(`/reliability/postmortems/${incidentId}`),
  runIntegrityCheck: () => api.post<Record<string, any>>('/reliability/integrity-check'),

  // Disaster Recovery
  listBackups: () => api.get<BackupRecord[]>('/disaster-recovery/backups'),
  triggerBackup: (data?: { backup_type?: string; storage_location?: string; retention_days?: number }) =>
    api.post<BackupRecord>('/disaster-recovery/backups', data || {}),
  listRestoreTests: () => api.get<RestoreVerificationTest[]>('/disaster-recovery/restore-tests'),
  runRestoreTest: (backupId: string, environment?: string) =>
    api.post<RestoreVerificationTest>('/disaster-recovery/restore-tests', { backup_id: backupId, environment }),
  listDRPlans: () => api.get<DRPlan[]>('/disaster-recovery/plans'),
  getDRPlan: (scenario: string) => api.get<DRPlan>(`/disaster-recovery/plans/${scenario}`),
  listDRDrills: () => api.get<DRDrill[]>('/disaster-recovery/drills'),
  runDRDrill: (scenario: string, initiatedBy: string, planName?: string) =>
    api.post<DRDrill>('/disaster-recovery/drills', { scenario, initiated_by: initiatedBy, plan_name: planName }),

  // Operations: Deployments & Flags
  listDeployments: () => api.get<DeploymentRecord[]>('/operations/deployments'),
  recordDeployment: (data: Partial<DeploymentRecord>) => api.post<DeploymentRecord>('/operations/deployments', data),
  evaluateAndRollback: (data: {
    current_version: string;
    target_version: string;
    error_rate_pct: number;
    p99_latency_ms: number;
    smoke_tests_passed: boolean;
    initiated_by: string;
    reason?: string;
  }) => api.post('/operations/rollbacks', data),
  listFeatureFlags: () => api.get<FeatureFlag[]>('/operations/feature-flags'),
  setFeatureFlag: (data: FeatureFlag) => api.post('/operations/feature-flags', data),
  evaluateFeatureFlag: (flagName: string, tenantId?: string, tenantTier?: string) =>
    api.post<{ flag_name: string; is_enabled: boolean }>('/operations/feature-flags/evaluate', {
      flag_name: flagName,
      tenant_id: tenantId,
      tenant_tier: tenantTier,
    }),
};
